kotlin/src/main/java/org/apache/zeppelin/kotlin/KotlinInterpreter.java

```diff
@@ -0,0 +1,191 @@
+import org.apache.zeppelin.interpreter.util.InterpreterOutputStream;
+import org.apache.zeppelin.kotlin.completion.KotlinCompleter;
+import org.apache.zeppelin.kotlin.context.KotlinReceiver;
+import org.apache.zeppelin.kotlin.reflect.KotlinFunctionInfo;
+import org.apache.zeppelin.kotlin.reflect.KotlinVariableInfo;
+import org.apache.zeppelin.kotlin.repl.KotlinRepl;
+import org.apache.zeppelin.kotlin.repl.building.KotlinReplProperties;
+import org.apache.zeppelin.scheduler.Job;
+
+public class KotlinInterpreter extends Interpreter {
+
+  private static Logger logger = LoggerFactory.getLogger(KotlinInterpreter.class);
+
+  private InterpreterOutputStream out;
+  private KotlinRepl interpreter;
+  private KotlinReplProperties replProperties;
+  private KotlinCompleter completer;
+
+  public KotlinInterpreter(Properties properties) {
+    super(properties);
+    replProperties = new KotlinReplProperties();
+
+    int maxResult = Integer.parseInt(
+        properties.getProperty("zeppelin.kotlin.maxResult", "1000"));
+
+    boolean shortenTypes = Boolean.parseBoolean(
+        properties.getProperty("zeppelin.kotlin.shortenTypes", "true"));
+    String imports = properties.getProperty("zeppelin.interpreter.localRepo", "");
+
+    completer = new KotlinCompleter();
+
+    replProperties
+        .receiver(new KotlinReceiver())
+        .maxResult(maxResult)
+        .codeOnLoad("")
+        .classPath(getImportClasspath(imports))
+        .shortenTypes(shortenTypes);
+  }
+
+  public KotlinReplProperties properties() {
+    return replProperties;
+  }
+
+  @Override
+  public void open() throws InterpreterException {
+    interpreter = new KotlinRepl(replProperties);
+
+    completer.setCtx(interpreter.getKotlinContext());
+    out = new InterpreterOutputStream(logger);
+  }
+
+  @Override
+  public void close() {
+
+  }
+
+  @Override
+  public InterpreterResult interpret(String code,
+                                     InterpreterContext context) throws InterpreterException{
+    // saving job's running thread for cancelling
+    Job<?> runningJob = getRunningJob(context.getParagraphId());
+    if (runningJob != null) {
+      runningJob.info().put("CURRENT_THREAD", Thread.currentThread());
+    }
+
+    return runWithOutput(code, context.out);
+  }
+
+  @Override
+  public void cancel(InterpreterContext context) throws InterpreterException {
+    Job<?> runningJob = getRunningJob(context.getParagraphId());
+    if (runningJob != null) {
+      Map<String, Object> info = runningJob.info();
+      Object object = info.get("CURRENT_THREAD");
+      if (object instanceof Thread) {
+        try {
+          Thread t = (Thread) object;
+          t.interrupt();
+        } catch (Throwable t) {
+          logger.error("Failed to cancel script: " + t, t);
+        }
+      }
+    }
+  }
+
+  @Override
+  public FormType getFormType() throws InterpreterException {
+    return FormType.NATIVE;
```

Comment:
Kotlin doesn't support dynamic forms ? If yes, it should be FormType.SIMPE

Reply:
Kotlin uses `$` for string interpolation, so dynamic forms conflict with this notation. 
Though Kotlin Spark Interpreter binds `ZeppelinContext` and can work with it.
Should I add this feature to base Kotlin Interpreter as well?